paper/src/main/java/com/denizenscript/denizen/paper/events/TargetHitScriptEvent.java

```diff
@@ -0,0 +1,98 @@
+
+import com.denizenscript.denizen.events.BukkitScriptEvent;
+import com.denizenscript.denizen.objects.EntityTag;
+import com.denizenscript.denizen.objects.LocationTag;
+import com.denizenscript.denizen.utilities.implementation.BukkitScriptEntryData;
+import com.denizenscript.denizencore.objects.ObjectTag;
+import com.denizenscript.denizencore.objects.core.ElementTag;
+import com.denizenscript.denizencore.scripts.ScriptEntryData;
+import io.papermc.paper.event.block.TargetHitEvent;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+
+
+public class TargetHitScriptEvent extends BukkitScriptEvent implements Listener {
+
+    // <--[event]
+    // @Events
+    // targetblock hit
+    //
+    // @Location true
+    //
+    // @Group Paper
+    //
+    // @Plugin Paper
+    //
+    // @Triggers when a target block is hit by a projectile such as an arrow.
+    //
+    // @Context
+    // <context.projectile> returns an EntityTag of the projectile.
+    // <context.shooter> returns a EntityTag of the shooter of the projectile.
+    // <context.hit_block> returns a LocationTag of the block
+    // <context.hit_face> returns a LocationTag of the face
+    // <context.strength> returns a ElementTag of the emitted redstone strength
+    //
+    // @Player when the shooter is a player.
+    //
+    // @NPC when the shooter is a npc.
+    // -->
+
+    public TargetHitScriptEvent() {
+    }
+
+    public TargetHitEvent event;
+    public LocationTag hitBlock;
+    public EntityTag projectile;
+    public EntityTag shooter;
+    public ElementTag strength;
+
+    @Override
+    public boolean couldMatch(ScriptPath path) {
+        return path.eventLower.matches("targetblock hit");
+    }
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        if (!runInCheck(path, hitBlock)) {
+            return false;
+        }
+        return super.matches(path);
+    }
+
+    @Override
+    public ObjectTag getContext(String name) {
+        return switch (name) {
+            case "projectile" -> projectile.getDenizenObject();
+            case "hit_block" -> hitBlock;
+            case "hit_face" -> event.getHitBlockFace() != null ? new LocationTag(event.getHitBlockFace().getDirection()) : null;
+            case "shooter" -> shooter != null ? shooter.getDenizenObject() : null;
+            case "strength" -> strength;
+            default -> super.getContext(name);
+        };
+    }
+
+    @Override
+    public ScriptEntryData getScriptEntryData() {
+        return new BukkitScriptEntryData(shooter);
+    }
+
+    @EventHandler
+    public void onProjectileHit(TargetHitEvent event) {
+        this.event = event;
+        projectile = new EntityTag(event.getEntity());
+
+        // Additional checks for some rare edge-cases
+        if (projectile.getLocation() == null) {
+            return;
+        }
+        if (Double.isNaN(projectile.getLocation().getDirection().normalize().getX())) {
```

Comment:
This is clearly copypasted from the projectile hit event - is it even relevant to have here?

Reply:
Relatedly - is this actually a unique event? ie is it not already covered by the existing projectile hit event?

Reply:
Glance at source code `public class TargetHitEvent extends ProjectileHitEvent` indicates it's not unique

Reply:
I don't see a reason why this should be needed. I copied it from the ProjectileHitEvent. In paper the Location of an entity is marked with @NotNull so I removed those checks

Reply:
> Glance at source code `public class TargetHitEvent extends ProjectileHitEvent` indicates it's not unique

It has a different HandlerList so it's a different event